### ISSUE TITLE
Do not pass lambdas to the backend in case of basic GroupBy aggregation

### ISSUE THREAD

**Issue by dchigarev** [modin-project/modin#3197]
Current GroupBy implementation is still passing python lambdas to the backend to do simple aggregation (for example `mean`):
https://github.com/modin-project/modin/blob/89f6cdde56abf5dcf9561ea6721fb31f84a855d5/modin/pandas/groupby.py#L128-L129
These lambdas are not processable by non-python engines, and so even if the backend does support particular aggregation it can't be executed because it's passed in a python function format. At the moment, all of those lambdas are being passed into a single method `QueryCompiler.groupby_agg`.

**Comment by dchigarev** [comment]
@devin-petersohn what do you think would be the most appropriate way to resolve the lambdas problem: 
1. We may continue to pass the aggregation method into a single `groupby_agg` backend function but as a string function name.
2. We may create separate query compiler methods for all of them, like we already did for the GroupBy reduce aggregations (`groupby_sum, groupby_count...` and we'd also add `groupby_mean, groupby_std, groupby_median...`)

**Comment by devin-petersohn** [comment]
+1 thanks @dchigarev, this is not efficient or ideal. I prefer to have some more explicit control in the query compiler. The hard part is that we end up with a huge bloat in the query compiler API if we explicitly have each API defined individually.

Maybe we need a better `GroupByFunction` to avoid the bloat? It is probably complementary with option 2 above.

**Comment by dchigarev** [comment]
@devin-petersohn 
Not sure that I understand what you mean under "better `GroupByFunction`", is it about those template functions that we store in [`data_management/functions`](https://github.com/modin-project/modin/tree/master/modin/data_management/functions)? If so, then I don't get how we could avoid extraordinary size of the query compiler API with a better `GroupByFunction`. The way I see your wish is the following:
```python
groupby_mean = GroupByFunction.register(pandas.GroupBy.mean)
groupby_nunique = GroupByFunction.register(pandas.GroupBy.nunique)
groupby_std = GroupByFunction.register(pandas.GroupBy.std)
...
```
In this approach, the problem is still unsolved: query compilers have a huge mixed API and potentially enormous size. The problem of having a huge group of methods related to the different high-level objects, but located in a single query compiler class is not groupby specific: we have the same issue with the [`resample_*`](https://github.com/modin-project/modin/blob/0bc409d6ae294986db5f7db19d082622544fb122/modin/backends/pandas/query_compiler.py#L862-L1022), [`rolling_*`](https://github.com/modin-project/modin/blob/0bc409d6ae294986db5f7db19d082622544fb122/modin/backends/pandas/query_compiler.py#L1024-L1152), [`str_*`](https://github.com/modin-project/modin/blob/0bc409d6ae294986db5f7db19d082622544fb122/modin/backends/pandas/query_compiler.py#L1357-L1402), and [`dt_*`](https://github.com/modin-project/modin/blob/0bc409d6ae294986db5f7db19d082622544fb122/modin/backends/pandas/query_compiler.py#L1426-L1485) methods. Bringing `groupby_*` methods to the query compiler will no create any new issues, it'll just highlight the existed one.

In a long term, I think, we should start thinking about splitting query compiler somehow. I think we could move the logic of actual `object_*` functions into different object-specific classes, something like: `QueryCompilerGroupByMethods`, `QueryCompilerWindowMethods`, and so on. Old query compiler `object_*` methods may be squeezed into just `groupby`, `resample`, `window` ..., that would be dispatchers of the `QueryCompiler$ObjectMethods` methods.

As a solution particularly for the #3197 issue I suggest expanding the query compiler's API with such methods:
```python
def groupby_mean(self, **kwargs):
    return self.groupby_agg(agg_func="mean", **kwargs)

def groupby_nunique(self, **kwargs):
    return self.groupby_agg(agg_func="nunique", **kwargs)
...
```
adding them (about 15 new methods) to the `BaseQueryCompiler` would allow us to not modify other query compilers at all. All of the groupby methods continue to work naturally via `groupby_agg`. At the same time, if a backend needs more control for some specific method it could just overwrite it.

**Comment by YarShev** [comment]
We likely want to keep QC API as clean as possible and all the implementation should go to a different location. We could move those implementation to `kernels` folder after #2957 is merged. Something like this:
```python
.pandas/
   query_compiler.py
   .kernels/
      .dataframe.py
      .groupby.py
      .resample.py
      .etc.
```

**Comment by YarShev** [comment]
Firstly I would like to see #3537 and #3562 merged into master. After that we can move forward more gradually and introduce more granular changes in the code and docs.

**Comment by dchigarev** [comment]
I think the problem we're discussing here (slice QC into several objects) is a bit different from the original one, IMO they could and should be solved independently. 

Splitting the Query Compiler is a right thing, but I think it's much more complicated, and what's important, much longer to implement, rather than passing backend-agnostic information about groupby aggregation (via function's string name or by calling a specific QC method) in the existing paradigm. The fact of defaulting to pandas for most of the OmniSci's groupby aggregation just because the type of aggregation is passed as python lambdas is hurting, and we would like to fix this as soon as possible. I propose we shake this issue a bit and try to come up with some solution that would conform to everyone. 

Just reminding the current status:
There [was proposed two options](https://github.com/modin-project/modin/issues/3197#issuecomment-870683374) of how we could provide to a backend non-python and non-pandas dependent information about an aggregation function:
1. Pass string function names to the `QC.groupby_agg` instead of python lambdas. The con of this approach is that it's hard to specialize the implementation of some concrete aggregation at the QC level (it would require a decent if-else tree to determine whether the passed operation is specialized)
2. Introduce new 15 QC methods named like `QC.groupby_{fn_aggregation_name}` and invoke aggregation through them (the same we do for `groupby.size`, `groupby.sum`, `groupby.count` ... and other MapReduce implemented fns). The con of this approach is that it will bloat Query Compiler's API.

Currently, there is an open PR (#3373) implementing the second approach. The PR [was postponed](https://github.com/modin-project/modin/pull/3373#pullrequestreview-739618708) until we will figure out how to avoid QC API bloating, however, we still haven't progressed much in it.

What I want from @modin-project/modin-core, @modin-project/modin-contributors, and @modin-project/modin-omnisci is to help to pick one of two proposed approaches or propose a new one. If we would agree on the second of the proposed approaches, I will make an effort to revive #3373 and make it ready for review (currently it's much outdated).

**Comment by devin-petersohn** [comment]
Many of our current groupby aggregations follow `mapreduce` or `treereduce` style of computation patterns. Many of the aggregations not yet implemented for `groupby` cannot follow this pattern. The new algebraic operator definitions in #3717 will help here a bit. `groupby` takes an operator (`reduce`, `map`, `treereduce`, `mask`, etc.). This is sufficient to support all of the `groupby` operators.

My vote is for (2) in @dchigarev's list here. The query compiler needs to be split, completely agree, but also agree with @dchigarev that it's orthogonal to the current discussion and can be done separately. Option (2) would allow us to individually define the groupby + operator pairing and let us optimize each independently. Other thoughts?

**Comment by devin-petersohn** [comment]
Yes @vnlitvinov let's do that.
